packages/pg-gateway/src/connection.ts

```diff
@@ -275,353 +279,301 @@ export default class PostgresConnection {
    * Inspired by https://github.com/brianc/node-postgres/blob/54eb0fa216aaccd727765641e7d1cf5da2bc483d/packages/pg-protocol/src/parser.ts#L91-L119
    */
   async handleData(data: Buffer) {
-    // Wrap in try-catch to prevent server from crashing during exceptions
     try {
-      this.mergeBuffer(data);
-
-      const bufferFullLength = this.bufferOffset + this.bufferLength;
-      let offset = this.bufferOffset;
-
-      // The initial message only has a 4 byte header containing the message length
-      // while all subsequent messages have a 5 byte header containing first a single
-      // byte code then a 4 byte message length
-      const codeLength = !this.hasStarted ? 0 : 1;
-      const headerLength = 4 + codeLength;
-
-      // If we have enough buffer to read the message header
-      while (offset + headerLength <= bufferFullLength) {
-        // The length passed in the message header
-        const length = this.buffer.readUInt32BE(offset + codeLength);
+      this.messageBuffer.mergeBuffer(data);
+      await this.messageBuffer.processMessages(
+        this.handleClientMessage.bind(this),
+        this.hasStarted,
+      );
+    } catch (err) {
+      console.error(err);
+    }
+  }
 
-        // The length passed in the message header does not include the first single
-        // byte code, so we account for it here
-        const fullMessageLength = codeLength + length;
+  async handleClientMessage(message: Buffer): Promise<void> {
+    this.reader.setBuffer(0, message);
 
-        // If we have enough buffer to read the entire message
-        if (offset + fullMessageLength <= bufferFullLength) {
-          // Create a Buffer view that points to the same memory as the main buffer,
-          // but crops it at the message boundaries
-          const messageData = this.buffer.subarray(offset, fullMessageLength);
+    this.socket.pause();
+    const messageSkip = await this.options.onMessage?.(message, this.state);
+    this.socket.resume();
 
-          // Process the message
-          await this.handleMessage(messageData);
+    if (messageSkip) {
+      return;
+    }
```

Comment:
Just a warning that we need to know if a startup message has occurred, even if `messageSkip` is `true`, otherwise future buffering will fail. This is because the first client message doesn't include a 1 byte code (this can be either a startup message or an ssl request) whereas every message after that does include this code. So if we skip processing during the initial message and don't mark the connection as `hasStarted = true`, the message buffer will incorrectly read the message length and in turn incorrectly handle buffering logic because is still thinks the first message hasn't been received yet.

Reply:
I'm now checking if the current message is the startup message if messageSkip is true. In this case I put `this.hasStarted` to true. 👍 
Fixed in https://github.com/supabase-community/pg-gateway/pull/1/commits/1a641dfaea1ad3483cf3521170a2e653ccaeefbb